src/main/scala/com/databricks/labs/overwatch/MultiWorkspaceDeployment.scala

```diff
@@ -0,0 +1,410 @@
+   *
+   * @param configDF
+   */
+  private def snapshotConfig(configDF: DataFrame) = {
+    var configWriteLocation = configDF.head().getAs(ConfigColumns.etl_storage_prefix.toString).toString
+    if (!configWriteLocation.startsWith("dbfs:") && !configWriteLocation.startsWith("s3") && !configWriteLocation.startsWith("abfss")) {
+      configWriteLocation = s"""dbfs:${configWriteLocation}"""
+    }
+    configDF
+      .withColumn("snapTS", lit(pipelineSnapTime.asTSString))
+      .withColumn("timestamp", lit(pipelineSnapTime.asUnixTimeMilli))
+      .withColumn("configFile", lit(configCsvPath))
+      .write.format("delta").mode("append").save(s"""${configWriteLocation}/report/configTable""")
+  }
+
+  /**
+   * Takes a snapshot of the validation result
+   *
+   * @param validationDF
+   * @param path
+   * @param reportName
+   */
+  private def snapShotValidation(validationDF: Dataset[DeploymentValidationReport], path: String, reportName: String): Unit = {
+    var validationPath = path
+    if (!path.startsWith("dbfs:") && !path.startsWith("s3") && !path.startsWith("abfss")) {
+      validationPath = s"""dbfs:${path}"""
+    }
+    validationDF
+      .withColumn("deployment_id", lit(deploymentId))
+      .withColumn("snapTS", lit(pipelineSnapTime.asTSString))
+      .withColumn("timestamp", lit(pipelineSnapTime.asUnixTimeMilli))
+      .write.format("delta").mode("append").save(s"""${validationPath}/report/${reportName}""")
+    println("Validation report has been saved to " + s"""${validationPath}/report/${reportName}""")
+  }
+
+  /**
+   * Saves the deployment report.
+   *
+   * @param validationDF
+   * @param path
+   * @param reportName
+   */
+  private def saveDeploymentReport(validationDF: Dataset[MultiWSDeploymentReport], path: String, reportName: String): Unit = {
+    var reportPath = path
+    if (!path.startsWith("dbfs:") && !path.startsWith("s3") && !path.startsWith("abfss")) {
+      reportPath = s"""dbfs:${path}"""
+    }
+    validationDF
+      .withColumn("snapTS", lit(pipelineSnapTime.asTSString))
+      .withColumn("timestamp", lit(pipelineSnapTime.asUnixTimeMilli))
+      .write.format("delta").mode("append").save(s"""${reportPath}/report/${reportName}""")
+  }
+
+  /**
+   * Performs the multi workspace deployment
+   *
+   * @param parallelism the number of threads which will run in parallel to perform the deployment
+   * @param zones       the zone can be Bronze,Silver or Gold by default it will be Bronze
+   */
+  def deploy(parallelism: Int = 4, zones: String = "Bronze"): Unit = {
+    val processingStartTime = System.currentTimeMillis();
+    setParallelism(parallelism)
+    println("ParallelismLevel :" + parallelism)
+    val deploymentValidation = DeploymentValidation
+    deploymentValidation.performMandatoryValidation(configCsvPath, parallelism, deploymentId)
+    val dataFrame = deploymentValidation.makeDataFrame(configCsvPath, deploymentId)
```

Comment:
this seems redundant -- why create a val for this? why not just

```
val dataframe = DeploymentValidation
    .performMandatoryValidation(configCsvPath, parallelism, deploymentId)
    .makeDataFrame(configCsvPath, deploymentId)
```

Reply:
performMandatoryValidation don't return anything so we cannot call .makeDataframe on top of .performMandatoryValidation

Reply:
this goes back to the bigger discussion. The CSV should be converted to a strongly typed DS once not for each validation. This should be an object level variable reusable across the entire scope. We can chat more.

Reply:
if we create the DS as part of class instantiation from companion object this won't be necessary here anymore at all.